libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -141,12 +166,73 @@ static void _destroyMqttConnection( _mqttConnection_t * pMqttConnection );
  */
 static IotMqttError_t _subscriptionCommon( IotMqttOperationType_t operation,
                                            IotMqttConnection_t mqttConnection,
+                                           IotMqttSubscribeSerializer_t serializeSubscription,
                                            const IotMqttSubscription_t * pSubscriptionList,
                                            size_t subscriptionCount,
                                            uint32_t flags,
                                            const IotMqttCallbackInfo_t * pCallbackInfo,
                                            IotMqttOperation_t * const pOperationReference );
 
+/**
+ * @cond DOXYGEN_IGNORE
+ * Doxygen should ignore this section.
+ *
+ * Declaration of local MQTT serializer override selectors
+ */
+#if IOT_MQTT_ENABLE_SERIALIZER_OVERRIDES == 1
+IOT_MQTT_SERIALIZER_OVERRIDE_SELECTOR(
```

Comment:
These macros are declaring and implementing functions?

It's difficult to figure out what's going on here... Also this would be difficult to debug since the debugger would jump to the parameterized macro.  You wouldn't be able to see what the function actually does.

Reply:
Yes, exactly.  Prior to this change, this code was duplicated __20 times__ across this library with the only differences being the function type and the serializer member, making the entire library difficult to read.  I considered making the serializer structure always present and initializing it during construction, but this would require more RAM even if you weren't overriding the serializer functions so I decided against it.  This is a deliberate decision to improve code readability at the possible expense (the macro code is very small) of making debugging slightly less convenient while not increasing our RAM use.